cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java

```diff
@@ -82,26 +83,40 @@ public TopicReplicaDistributionGoal(BalancingConstraint balancingConstraint) {
    * To avoid churns, we add a balance margin to the user specified rebalance threshold. e.g. when user sets the
    * threshold to be {@link BalancingConstraint#topicReplicaBalancePercentage()}, we use
    * ({@link BalancingConstraint#topicReplicaBalancePercentage()}-1)*{@link #BALANCE_MARGIN} instead.
+   *
+   * @param optimizationOptions Options to adjust balance percentage with margin in case goal optimization is triggered
+   * by goal violation detector.
    * @return the rebalance threshold with a margin.
    */
-  private double balancePercentageWithMargin() {
-    return (_balancingConstraint.topicReplicaBalancePercentage() - 1) * BALANCE_MARGIN;
+  private double balancePercentageWithMargin(OptimizationOptions optimizationOptions) {
+    double balancePercentage = optimizationOptions.isTriggeredByGoalViolation()
+                               ? _balancingConstraint.topicReplicaBalancePercentage()
+                                 * _balancingConstraint.goalViolationDistributionThresholdMultiplier()
+                               : _balancingConstraint.topicReplicaBalancePercentage();
+
```

Comment:
maybe we can change to
`_balancingConstraint.topicReplicaBalancePercentage() * (optimizationOptions.isTriggeredByGoalViolation() ? _balancingConstraint.goalViolationDistributionThresholdMultiplier() : 1)`
to reduce a call to `_balancingConstraint.topicReplicaBalancePercentage()`

Reply:
Replied above -- please let me know if I am missing something.